app/recurringjob/systembackup.go

```diff
@@ -0,0 +1,147 @@
+		"parameters":     job.parameters,
+		"executionCount": job.executionCount,
+		// system-backup-specific fields
+		"systemBackupName":   systemBackupName,
+		"volumeBackupPolicy": volumeBackupPolicy,
+	})
+
+	newJob := &SystemBackupJob{
+		Job:                job,
+		logger:             logger,
+		systemBackupName:   systemBackupName,
+		volumeBackupPolicy: longhorn.SystemBackupCreateVolumeBackupPolicy(volumeBackupPolicy),
+	}
+	return newJob, nil
+}
+
+func (job *SystemBackupJob) run() (err error) {
+	job.logger.Info("Starting system backup job")
+	defer func() {
+		if err != nil {
+			job.logger.WithError(err).Error("Failed to run system backup job")
+		} else {
+			job.logger.Info("Finished running system backup job")
+		}
+	}()
+
+	newSystemBackup := &longhorn.SystemBackup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      job.systemBackupName,
+			Namespace: job.namespace,
+			Labels: map[string]string{
+				types.GetRecurringJobLabelKey(types.LonghornLabelRecurringJob, string(longhorn.RecurringJobTypeSystemBackup)): job.name,
+			},
+		},
+		Spec: longhorn.SystemBackupSpec{
+			VolumeBackupPolicy: job.volumeBackupPolicy,
+		},
+	}
+
+	_, err = job.CreateSystemBackup(newSystemBackup)
+	if err != nil {
+		return err
+	}
+
+	finalStates := []longhorn.SystemBackupState{
+		longhorn.SystemBackupStateReady,
+		longhorn.SystemBackupStateError,
+	}
+	err = job.waitForSystemBackupToStates(finalStates)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (job *SystemBackupJob) cleanup() {
+	job.logger.Info("Cleaning up expired system backups")
+	defer job.logger.Info("Finished cleaning up expired system backups")
+
+	systemBackupList, err := job.ListSystemBackup()
+	if err != nil {
+		job.logger.WithError(err).Warn("Failed to list system backups")
+		return
+	}
+
+	expiredSystemBackups := filterExpiredItems(systemBackupsToNameWithTimestamps(systemBackupList), job.retain)
+	for _, systemBackupName := range expiredSystemBackups {
+		job.logger.Infof("Deleting system backup %v", systemBackupName)
+		err = job.DeleteSystemBackup(systemBackupName)
+		if err != nil {
+			job.logger.WithError(err).Warnf("Failed to delete system backup %v", systemBackupName)
+		}
+	}
+}
+
+func (job *SystemBackupJob) waitForSystemBackupToStates(expectedStates []longhorn.SystemBackupState) (err error) {
+	job.logger.Info("Waiting for system backup to complete")
+	defer func() {
+		if err != nil {
+			job.logger.WithError(err).Errorf("Failed to wait for system backup to reach state %v", expectedStates)
+		} else {
+			job.logger.Infof("System backup reached state %v", expectedStates)
+		}
+	}()
+
+	for {
+		systemBackup, err := job.GetSystemBackup(job.systemBackupName)
+		if err != nil {
+			return err
+		}
+
+		if lhutils.Contains(expectedStates, systemBackup.Status.State) {
+			return nil
+		}
+
+		job.logger.Infof("Waiting for system backup to reach state %v, current state is %v", expectedStates, systemBackup.Status.State)
+		time.Sleep(WaitInterval)
+	}
+}
```

Comment:
Is it possible an infinite loop? Should we consider adding a maximum loop count?

Reply:
I intentionally designed it this way because the time required for a system backup depends on the cluster, making it challenging to determine a reasonable maximum duration - especially if volume backups are involved.

Having this as a blocking operation shouldn't harm the recurring system backup. Even if it gets stuck, users can remove it and let the cronjob recreate it.

I prefer to wait and see if there's any feedback from the community. What do you think?

Reply:
Sure.